### PR TITLE
[BlockStoreLoop] Refactor the BlockStoreSteps so that it returns a enum result #330

### PR DESCRIPTION
**Pull request by fassadlr**

### CODE REVIEW

**Comment by bokobza** [comment]
I'll do this one later today.

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs [hunk 1]
```diff
@@ -35,21 +35,16 @@ internal void SetNextStep(BlockStoreLoopStep step)
...
-
             foreach (var step in this.steps)
             {
                 var stepResult = await step.ExecuteAsync(nextChainedBlock, cancellationToken, disposeMode);
-                if (stepResult.ShouldBreak || stepResult.ShouldContinue)
-                {
-                    result = stepResult;
-                    break;
-                }
+                if (stepResult == StepResult.Continue || stepResult == StepResult.Stop)
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs]: just a note fore the future - we intend to make a new code style policy that will include parentheses to be used as:

```
if ((stepResult == StepResult.Continue) || (stepResult == StepResult.Stop))
```

not a problem now as the new rules are not yet formed, but might be easier to comply already (see https://github.com/stratisproject/StratisBitcoinFullNode/issues/258) for more

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs [hunk 2]
```diff
@@ -78,41 +73,11 @@ protected BlockStoreLoopStep(BlockStoreLoop blockStoreLoop)
...
-        /// Will cause the caller to execute the next line of code.
-        /// </summary>
-        internal static BlockStoreLoopStepResult Next()
-        {
-            var result = new BlockStoreLoopStepResult();
-            return result;
-        }
+        Continue,
+        Next,
+        Stop,
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs]: Move comment from enum to each of the values. We always want the enum values to be commented - you have those comments already, but in wrong place.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs [hunk 3]
```diff
@@ -28,28 +28,32 @@ internal DownloadBlockStep(BlockStoreLoop blockStoreLoop)
...
+                        return StepResult.Next;
                 }
             }
 
-            return result;
+            return StepResult.Next;
         }
     }
+
+    public enum InnerStepResult
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: comments missing here - pls make sure each value is commented as well as the whole enum

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs [hunk 4]
```diff
@@ -94,29 +94,31 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop)
...
         }
 
         /// <summary>
         /// Break execution if:
-        ///     1: At the tip 
-        ///     2: Block is already in store or pending insertion
+        /// <list>
+        ///     <item>1: At the tip</item>
+        ///     <item>2: Block is already in store or pending insertion</item>
+        ///<list>
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs]: this should be `/// </list>` <- i.e. space and slash missing

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs [hunk 5]
```diff
@@ -28,28 +28,32 @@ internal DownloadBlockStep(BlockStoreLoop blockStoreLoop)
...
 
             while (!token.IsCancellationRequested)
             {
-                foreach (var item in context.Routine.ToList())
+                foreach (var innerStep in context.InnerSteps.ToList())
                 {
-                    var executionResult = await item.ExecuteAsync(context);
-                    if (executionResult.ShouldBreak)
-                        return result;
+                    var innerStepResult = await innerStep.ExecuteAsync(context);
```
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: this and above use of `var` is in violation of rule 10 of existing code style - https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/Documentation/coding-style.md

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs (hunk 4 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/ProcessPendingStorageStep.cs]: Fixed

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 3 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: Fixed

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 5 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: Fixed

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs (hunk 2 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs]: Fixed

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs (hunk 1 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoopStep.cs]: Fixed

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 3 above)
Comment (Aprogiena) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: I still can't see any comment here :)

Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: Sorry! It has been added now :)

**Comment by bokobza** [comment]
ah it's fine @Aprogiena, thanks for doing it.

**Pull request merged**